snapshots/manager.go

```diff
@@ -0,0 +1,223 @@
+package snapshots
+
+import (
+	"bytes"
+	"crypto/sha256"
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"sync"
+	"time"
+)
+
+const (
+	opNone     operation = ""
+	opSnapshot operation = "snapshot"
+	opPrune    operation = "prune"
+	opRestore  operation = "restore"
+
+	chunkBufferSize = 4
+)
+
+// operation represents a Manager operation. Only one operation can be in progress at a time.
+type operation string
+
+// Manager manages snapshot and restore operations for an app, making sure only a single
+// long-running operation is in progress at any given time, and provides convenience methods
+// mirroring the ABCI interface.
+type Manager struct {
+	store  *Store
+	target Snapshotter
+
+	mtx            sync.Mutex
+	operation      operation
+	chRestore      chan<- io.ReadCloser
+	chRestoreDone  <-chan error
+	restorePending [][]byte // pending chunk hashes
+}
+
+// NewManager creates a new manager.
+func NewManager(store *Store, target Snapshotter) *Manager {
+	return &Manager{
+		store:  store,
+		target: target,
+	}
+}
+
+// begin starts an operation, or errors if one is in progress. It manages the mutex itself.
+func (m *Manager) begin(op operation) error {
+	m.mtx.Lock()
+	defer m.mtx.Unlock()
+	return m.beginLocked(op)
+}
+
+// beginLocked begins an operation while already holding the mutex.
+func (m *Manager) beginLocked(op operation) error {
+	if op == opNone {
+		return errors.New("can't begin a none operation")
```

Comment:
ditto, this is also an internal logic error, would be good to mark it with a specific type.